### PR TITLE
discovery: coalesce identical SD configurations

### LINKED ISSUE #2191: Consider coalescing identical SD definitions
There may be cases where users have multiple different scrape configs that share a SD configuration, but where other settings such as relabelling are different.

We have been slowly adding the ability to filter down at the SD level in an ad-hoc way for each SD, but given the number of things that a Prometheus is likely monitoring I think there's another potential approach to keeping load on the SD sane.

If there's two identical SD configs, let's only instantiate it once and then feed it into each required scrape config. This means a naieve config will have constant load on SD relative to the number of scrape configs, and avoid a good chunk of users having to worry about SD performance.

### PR DESCRIPTION
**Pull request by simonpasquier**
Closes #2191.

For now, it is still early stages but it looks like it is working in my environment. I'm going to test with lots of identical k8s configs and check if we see noticeable changes on Prometheus/k8s sides.
cc @brian-brazil @krasi-georgiev @fabxc.

### CODE REVIEW

**Comment by krasi-georgiev** [comment]
I will let @Gouthamve have a look at this one if he has time  as a present for his return :)

I am tackling some races in tsdb.

File: discovery/manager.go [hunk 1]
```diff
@@ -59,8 +60,18 @@ type poolKey struct {
 	provider string
 }
 
+type provider struct {
+	name   string
+	d      Discoverer
+	sets   []string
```
Comment (cstyan) [inline discovery/manager.go]: it's not entirely clear to me what sets are supposed to be

File: discovery/manager.go [hunk 2]
```diff
@@ -203,85 +222,134 @@ func (m *Manager) allGroups() map[string][]*targetgroup.Group {
...
-	providers := map[string]Discoverer{}
-
-	app := func(mech string, i int, tp Discoverer) {
-		providers[fmt.Sprintf("%s/%d", mech, i)] = tp
+func (m *Manager) registerProviders(cfg sd_config.ServiceDiscoveryConfig, setName string) {
+	register := func(cfg interface{}) bool {
+		for _, p := range m.providers {
+			if reflect.DeepEqual(cfg, p.config) {
+				p.sets = append(p.sets, setName)
+				//level.Debug(m.logger).Log("msg", "Registering new set for provider", "provider", p.name, "sets", fmt.Sprintf("%v", p.sets))
```
Comment (cstyan) [inline discovery/manager.go]: did you want to keep these log lines?

Comment (simonpasquier) [reply discovery/manager.go]: Thanks for spotting!

File: discovery/manager.go (hunk 1 above)
Comment (simonpasquier) [reply discovery/manager.go]: I used `sets` because `poolKey` has `setName` but I agree that the naming can be improved.

Comment (cstyan) [reply discovery/manager.go]: okay so then `setName` is either scrape config `jobName` or the md5 checksum of the AlertmanagerConfig as JSON

Comment (simonpasquier) [reply discovery/manager.go]: Correct

**Comment by brancz** [comment]
lgtm

**Comment by solidspark** [comment]
+1, we could use this.

**Comment by grobie** [comment]
Lgtm, but I think we should fix #4124 first.

File: discovery/manager.go [hunk 3]
```diff
@@ -210,85 +234,132 @@ func (m *Manager) allGroups() map[string][]*targetgroup.Group {
...
+			config: cfg,
+			subs:   []string{setName},
+		}
+		m.providers = append(m.providers, &provider)
 	}
 
-	for i, c := range cfg.DNSSDConfigs {
-		app("dns", i, dns.NewDiscovery(*c, log.With(m.logger, "discovery", "dns")))
+	for _, c := range cfg.DNSSDConfigs {
+		if register(c) {
```
Comment (brian-brazil) [inline discovery/manager.go]: I think it might be clearer to put all this into one function, rather than splitting across two

**Comment by simonpasquier** [comment]
@brian-brazil I've added `prometheus_sd_failed_configs`, a gauge metric with a `name` label that is either `scrape` or `notify`.

**Comment by brian-brazil** [comment]
That's not a great metric, I'd suggest an unlabelled counter.

**Comment by simonpasquier** [comment]
I thought about having a counter first but I didn't see how it could be useful for alerting as the value wouldn't change after the configuration has been fixed.

**Comment by brian-brazil** [comment]
The problem with what's here is that once there's a failure it'll stick. In general with SD we expose failure count metrics, so I think we should stick with that.

**Comment by brian-brazil** [comment]
:+1:

File: discovery/manager.go [hunk 4]
```diff
@@ -210,85 +248,96 @@ func (m *Manager) allGroups() map[string][]*targetgroup.Group {
...
+		for _, p := range m.providers {
+			if reflect.DeepEqual(cfg, p.config) {
+				p.subs = append(p.subs, setName)
+				return
+			}
+		}
+
+		d, err := df()
+		if err != nil {
+			level.Error(m.logger).Log("msg", "Cannot create service discovery", "err", err, "type", t)
```
Comment (krasi-georgiev) [inline discovery/manager.go]: instead of passing `t string` can't we use  something like `reflect.TypeOf(cfg).Elem()`

File: discovery/manager.go [hunk 5]
```diff
@@ -210,85 +248,96 @@ func (m *Manager) allGroups() map[string][]*targetgroup.Group {
...
+func (m *Manager) registerProviders(cfg sd_config.ServiceDiscoveryConfig, setName string) {
+	add := func(t string, cfg interface{}, df func() (Discoverer, error)) {
+		for _, p := range m.providers {
+			if reflect.DeepEqual(cfg, p.config) {
+				p.subs = append(p.subs, setName)
+				return
+			}
+		}
+
+		d, err := df()
```
Comment (krasi-georgiev) [inline discovery/manager.go]: wouldn't be better to call this something more descriptive like: `NewDiscoverer`

File: discovery/manager.go [hunk 6]
```diff
@@ -146,7 +181,9 @@ func (m *Manager) runProvider(ctx context.Context, poolKey poolKey, updates chan
 			if !ok {
 				return
 			}
-			m.updateGroup(poolKey, tgs)
+			for _, s := range p.subs {
+				m.updateGroup(poolKey{setName: s, provider: p.name}, tgs)
```
Comment (krasi-georgiev) [inline discovery/manager.go]: why not use the `p *provider` as a pool key directly?  I used a seperate struct just to make it unique enough, but now with the `type provider struct` maybe we can use it directly and remove the `poolKey type` entirely.

Comment (simonpasquier) [reply discovery/manager.go]: Hmm we still need to dispatch the target groups from 1 provider to N jobs at some point.

**Comment by simonpasquier** [comment]
I didn't test specifically with `-race` but as you said, this code tends to be racy so will do.

**Comment by brian-brazil** [comment]
You've got a conflict

**Comment by brian-brazil** [comment]
Anyone have any final comments before I merge?

**Comment by brian-brazil** [comment]
Those issues are all from code introduced after this PR, and I don't think
they should be blocking it. Rather we should be thinking about reverting
them as they're causing issues.

**Comment by krasi-georgiev** [comment]
> Those issues are all from code introduced after this PR, and I don't think
they should be blocking it. Rather we should be thinking about reverting
them as they're causing issues.

@brian-brazil I am talking about fixing  #4124 first. Which is the one you are referring to?

@grobie  there are 2 issues
1. exhaustion of open HTTP2 streams
   -  this PR mitigates the problem, but doesn't guarantee that you won't hit the exhaustion. A proper fix would be to revert the  http2  package, before https://github.com/golang/net/commit/1c05540f6879653db88113bc4a2b70aec4bd491f and wait for a fix upstream.
2. Processing the scrape reloading in serial which takes long to process so on systems with constant changes the reloading can never keep up.
   - fixed in #4526

**Comment by krasi-georgiev** [comment]
@brian-brazil  any objections if  I open a PR to revert the http package and possibly add a test to catch exhaustion of open HTTP2 streams regression.

govendor fetch golang.org/x/net/...@167db312be3be3d239e24dbe0ab099ab5b9aa783

**Comment by brian-brazil** [comment]
> Which is the one you are referring to?

I'd have to dig it up, but I have gotten the impression that there have been a series of changes each of which introduced new issues. If we can't get that under control, maybe we should be considering a revert.

**Comment by krasi-georgiev** [comment]
the only change that I am aware of is for the  throttling in https://github.com/prometheus/prometheus/pull/4523 (merged) and a small fix in  https://github.com/prometheus/prometheus/pull/4556 (waiting for a review)

I would like to move forward with the fix and can do a RC right after that, but if you prefer to revert that is fine by me.

**Comment by brian-brazil** [comment]
Those changes broke static_configs per #4551, if we're missing breakage to such basic functionality then what else are we missing?

**Comment by krasi-georgiev** [comment]
LGTM for this PR , thanks @simonpasquier

**Comment by brian-brazil** [comment]
> #4556 confirms to fix the behaviour, but as I said if you prefer to revert that is fine.

I'm not saying we have to revert, but we should be taking due care with such key and relatively intricate code. It seems every time someone touches it that it breaks.

**Pull request merged**